sway/input/seatop_default.c

```diff
@@ -452,38 +452,49 @@ static void handle_button(struct sway_seat *seat, uint32_t time_msec,
 		}
 	}
 
+	// Handle changing focus when clicking on a container
+	if (cont && state == WLR_BUTTON_PRESSED) {
+		// Default case: focus the container that was just clicked.
+		node = &cont->node;
+
+		// If the container is a tab/stacked container and the click happened
+		// on a tab, switch to the tab. If the tab contents were already
+		// focused, focus the tab container itself. If the tab container was
+		// already focused, cycle back to focusing the tab contents.
+		if (on_titlebar) {
+			struct sway_container *focus = seat_get_focused_container(seat);
+			if (focus == cont || !container_has_ancestor(focus, cont)) {
+				node = seat_get_focus_inactive(seat, &cont->node);
+			}
+		}
+
+		seat_set_focus(seat, node);
+	}
+
 	// Handle moving a tiling container
 	if (config->tiling_drag && (mod_pressed || on_titlebar) &&
 			state == WLR_BUTTON_PRESSED && !is_floating_or_child &&
-			cont && cont->fullscreen_mode == FULLSCREEN_NONE) {
-		struct sway_container *focus = seat_get_focused_container(seat);
-		bool focused = focus == cont || container_has_ancestor(focus, cont);
-		if (on_titlebar && !focused) {
-			node = seat_get_focus_inactive(seat, &cont->node);
-			seat_set_focus(seat, node);
-		}
-
-		// If moving a container by it's title bar, use a threshold for the drag
+			cont && cont->fullscreen_mode == FULLSCREEN_NONE &&
+			cont == seat_get_focused_container(seat)) {
```

Comment:
This condition is new. It's meant to prevent a "weird" UI state where the current tab container is focused, and a drag is initiated. The new block above will have switched the focus to the child, so now the tab container looks "inactive" but still responds to the drag. This isn't super intuitive or obvious, so I've disallowed the case.

Reply:
While testing, I kept focusing the parent and then dragging, which is what this prevents. This isn't functionality that I use though so I'm not sure how common that order would be.

Maybe it would actually be better to always focus `cont` here and then do the focus inactive handling in `finalize_move` in `seatop_move_tiling.c`?

Reply:
I like that idea a lot, actually. It'd add some complexity to the logic, since `seatop_begin_move_tiling` would have to now take a `struct sway_container *` to restore focus to at the end, iff there was no move... but the UX certainly would be much better.